clang/lib/DPCT/APINamesCUSPARSE.inc

```diff
@@ -153,8 +153,8 @@ ASSIGNABLE_FACTORY(ASSIGN_FACTORY_ENTRY(
     "cusparseCreateCsr", DEREF(0),
     CALL("std::make_shared<" + MapNames::getDpctNamespace() +
              "sparse::sparse_matrix_desc>",
-         ARG(1), ARG(2), ARG(3), ARG(4), ARG(5), ARG(6), ARG(7), ARG(8), ARG(9),
-         ARG(10),
+         ARG(1), ARG(2), ARG(3), NULLPTR_OR_ARG(4), NULLPTR_OR_ARG(5),
+         NULLPTR_OR_ARG(6), ARG(7), ARG(8), ARG(9), ARG(10),
```

Comment:
Assuming `ARG(4)` is for row-pointer array, shouldn't it be the following?
```suggestion
         ARG(1), ARG(2), ARG(3), ARG(4), NULLPTR_OR_ARG(5),
         NULLPTR_OR_ARG(6), ARG(7), ARG(8), ARG(9), ARG(10),
```
The size of the matrices are always known _a priori_, which means we know the length of the row-pointer array (`= nRows + 1`).

oneMKL sparse BLAS SYCL APIs do not support modifying arrays inside a matrix handle once they have been set, see "User agreements (2)" in the [documentation](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/sparse-blas-matrix-handle-contract-between-user.html):
> The User agrees not to modify the data arrays directly while they are in a sparse::matrix_handle_t.

Reply:
Yes, I agree this size of the matrix is known here.
But cusparse support to pass nullptr when creating a `cusparseSpMatDescr_t` and then use some set API to set those pointers.
So the cuda code may like below:
```
cusparseSpMatDescr_t* spMatDescr;
cusparseCreateCsr(&spMatDescr, rows, cols, nnz, nullptr, nullptr, nullptr, ...);
cudaMalloc(...);
cusparseCsrSetPointers(spMatDescr, ...);
```
SYCLomatic cannot move the malloc before constructing descriptor (we cannot change the original code logic).
Do you have any suggestion to support the above case?

Reply:
And I checked the example `sparse_matmat_usm.cpp` in MKL package in 2023.2
In this file, matrix handle is set data twice:
```
249        auto ev_setC = oneapi::mkl::sparse::set_csr_data(main_queue, C, c_nrows, c_ncols, c_index, c_rowptr,
250                                                        (intType *)nullptr, (fpType *)nullptr, {});
...
333        ev_setC = oneapi::mkl::sparse::set_csr_data(main_queue, C, c_nrows, c_ncols, c_index, c_rowptr, c_colind, c_values, {ev3_1});
```
Does it violate the "User agreements (2)"?

Reply:
To be clearer, _once the pointers have been set_ in oneMKL's `matrix_handle_t` type and they are no longer `nullptr`, they should not be switched out. It's okay to create the handle with `nullptr`s first, but once you set them to some pointer containing data, you should not switch it out with some other array -- I believe that is not allowed. Once you set the `_temp_row_ptr` in `matrix_handle_t`, switching it out with `_row_ptr` even if you are copying it could lead to unpredictable behavior.

Reply:
So here, I think it might be okay to keep `NULLPTR_OR_ARG(4)`.

Reply:
OK. But there is still one issue left.
The cusparse allows to create `cusparseSpMatDescr_t` with nullptr of `row_ptr`, while the `matrix_handle_t` does not allow it.
```
  1 cusparseSpMatDescr_t* matC;
  2 cusparseCreateCsr(&matC, rows, cols, nnz, NULL, NULL, NULL, ...); // MKL need row_ptr ready here
  3 cusparseSpGEMM_workEstimation(...);
  4 cusparseSpGEMM_workEstimation(...);
  5 cusparseSpGEMM_compute(...);
  6 cusparseSpGEMM_compute(...);
  7 cusparseSpMatGetSize(...);
  8 cudaMalloc(&cols, ...);
  9 cudaMalloc(&values, ...);
 10 cudaMalloc(&row_ptr, ...);
 11 cusparseCsrSetPointers(matC, ..., cols, values, row_ptr); // cusparse can set row_ptr here
 12 cusparseSpGEMM_copy(...);
```
Since `matrix_handle_t` does not allow switch the pointer `row_ptr` at line 11, can we create a totally new `matrix_handle_t` at line 11 and use it directly in line 12?

Reply:
So I see what's happening now: There's a difference between how cuSPARSE finalizes things and how oneMKL does it. In case of cuSPARSE and oneMKL, the work estimation stages match up, but the compute and finalize/spgemm_copy stages don't exactly match. The problem is that oneMKL finalizes/fills up the `row_ptr` of the C matrix in the compute stage itself, and finalizes/fills up the `cols`, `values` arrays of the C matrix in the finalize stage, but on the other hand, cuSPARSE finalizes/copies/fills up the C matrix in a single step in the end in their copy step. The steps for using oneMKL are:
```
sparse::matrix_handle_t matC;
malloc_device(row_ptr, sizeof(IntType) * (nrows+1));
sparse::set_csr_data(matC, rows, cols, row_ptr, NULL, NULL, ...); // row_ptr is necessary here
sparse::matmat(..., matmat_request::get_work_estimation_buf_size);
// Allocate tempBuffer1
sparse::matmat(..., matmat_request::work_estimation);
sparse::matmat(..., matmat_request::get_compute_buf_size);
sparse::matmat(..., matmat_request::compute);
sparse::matmat(..., matmat_request::get_nnz);
malloc_device(&cols, sizeof(IntType) * nnz);
malloc_device(&values, sizeof(DataType) * nnz);
sparse::set_csr_data(matC, rows, cols, row_ptr, cols, values, ...); // row_ptr is repeated here
sparse::matmat(..., matmat_request::finalize);
```

Reply:
I don't know how much control you have over the mapping and porting over in the DPCT tool, but note that we might not be able to change how we do these steps in oneMKL. The oneMKL `matmat` [documentation](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/oneapi-mkl-sparse-matmat.html) says in a table there:
<html>
<body>
<!--StartFragment-->

compute/compute_structure | do internal products for computing the C matrix including the calculation of size of C matrix data and filling the row pointer array for C.
-- | --


<!--EndFragment-->
</body>
</html>

Reply:
Then I think the current solution is: DPCT emit warning and throw exception for NULL row_ptr. If oneMKL changes the behavior in the future, we can update the helper function.

Reply:
I updated the PR.